frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -20,15 +20,50 @@ limitations under the License.
 namespace P4 {
 
 std::vector<const IR::IDeclaration*>*
-ResolutionContext::resolve(IR::ID name, P4::ResolutionType type, bool forwardOK) const {
+ResolutionContext::resolve(
+    IR::ID name, P4::ResolutionType type, bool forwardOK, const Visitor::Context* context) const {
     static std::vector<const IR::IDeclaration*> empty;
 
     std::vector<const IR::INamespace*> toTry(stack);
     toTry.insert(toTry.end(), globals.begin(), globals.end());
 
     for (auto it = toTry.rbegin(); it != toTry.rend(); ++it) {
         const IR::INamespace* current = *it;
-        LOG3("Trying to resolve in " << current->toString());
+        LOG3("Trying to resolve in " << current->toString() << " " << dbp(current));
+
+        // Return true if a declaration satisfies several criteria:
+        // - it comes before the name in the source
+        // - it is not enclosed within the namespace itself (unless the namespace isRecursive).
+        std::function<bool(const IR::IDeclaration*)> filter =
+                [name, forwardOK, context](const IR::IDeclaration* d) {
+            if (name.srcInfo.isValid() && !forwardOK) {
+                Util::SourceInfo nsi = name.srcInfo;
+                Util::SourceInfo dsi = d->getNode()->srcInfo;
+                bool before = dsi <= nsi;
+                LOG3("\tPosition test for " << dbp(d->getNode()) << ":"
+                     << dsi << "<=" << nsi << "=" << before);
+                if (!before) return false;
+            }
+
+            auto ns = d->to<IR::INamespace>();
+            if (!ns || !ns->isRecursive()) {
+                // Check to see whether d is a parent of the current
+                // node in the IR DAG; in this case we cannot resolve
+                // to it.  An example would be: table t { key = { t: exact }}
+                // However, we allow "recursive" namespaces resolve to identifiers
+                // within their context; this is necessary for example to allow
+                // transitions from a state to itself.
```

Comment:
This seems like `t` should resolve to the table (not to whatever t might be in the enclosing scope), which would then be an error (can't use a table as a key expression)...